### ISSUE TITLE
threading, multiple models in the same program.

### ISSUE THREAD

**Issue by iceisfun** [ultralytics/ultralytics#2273]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

We have two threads doing prediction with different models in the same python program.

it seems like model.predict must be in a Locked section to not get crashes and cuda errors, is there some kind of global shared state between instances of model?

here is a example that produces this crash

```
import threading
import requests
import numpy as np
import cv2
import time

response = requests.get("https://ultralytics.com/images/bus.jpg")
buf = np.asarray(bytearray(response.content), dtype=np.uint8)
img = cv2.imdecode(buf, cv2.IMREAD_COLOR)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

from ultralytics import YOLO
mtx=threading.Lock()
def threadmain():
    model = YOLO("yolov8n.pt")

    while True:
        # works, blocks other threads
        #with mtx:
        #    results = model(img)

        # crash path
        results = model(img)
        for item in results:
            print(item.boxes)

ts=[]
for x in range(4):
    t=threading.Thread(target=threadmain)
    t.start()
    ts.append(t)

for t in ts:
    t.join()



```

and here is a partial paste of the resulting crash

```
Exception in thread Thread-4 (threadmain):
Traceback (most recent call last):
  File "/usr/src/ultralytics/ultralytics/yolo/engine/predictor.py", line 221, in stream_inference
    preds = self.model(im, augment=self.args.augment, visualize=visualize)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
Exception in thread Exception in thread Thread-1 (threadmain):
Thread-2 (threadmain):
Traceback (most recent call last):
  File "/usr/src/ultralytics/ultralytics/yolo/engine/predictor.py", line 221, in stream_inference
Traceback (most recent call last):
    return forward_call(*args, **kwargs)
  File "/usr/src/ultralytics/ultralytics/yolo/engine/predictor.py", line 221, in stream_inference
  File "/usr/src/ultralytics/ultralytics/nn/autobackend.py", line 313, in forward
    preds = self.model(im, augment=self.args.augment, visualize=visualize)    preds = self.model(im, augment=self.args.augment, visualize=visualize)

  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
      File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
y = self.model(im, augment=augment, visualize=visualize) if augment or visualize else self.model(im)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/src/ultralytics/ultralytics/nn/autobackend.py", line 313, in forward
    return forward_call(*args, **kwargs)
Exception in thread     y = self.model(im, augment=augment, visualize=visualize) if augment or visualize else self.model(im)
      File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 205, in forward
Thread-3 (threadmain):
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
return forward_call(*args, **kwargs)
  File "/usr/src/ultralytics/ultralytics/nn/autobackend.py", line 313, in forward
        self.run()return self._forward_once(x, profile, visualize)  # single-scale inference, train
    return forward_call(*args, **kwargs)

  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 205, in forward
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 58, in _forward_once
    y = self.model(im, augment=augment, visualize=visualize) if augment or visualize else self.model(im)
  File "/opt/conda/lib/python3.10/threading.py", line 953, in run
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
            return self._forward_once(x, profile, visualize)  # single-scale inference, train
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 58, in _forward_once
x = m(x)  # run
        return forward_call(*args, **kwargs)
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 205, in forward
self._target(*self._args, **self._kwargs)
x = m(x)  # run  File "/project/cuda6.py", line 26, in threadmain
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl

      File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
return self._forward_once(x, profile, visualize)  # single-scale inference, train
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 58, in _forward_once
            return forward_call(*args, **kwargs)return forward_call(*args, **kwargs)
  File "/usr/src/ultralytics/ultralytics/nn/modules.py", line 409, in forward
print(item.boxes)

  File "/usr/src/ultralytics/ultralytics/nn/modules.py", line 40, in forward_fuse
    x = m(x)  # run
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return torch.cat(x, self.d)
RuntimeError:   File "/usr/src/ultralytics/ultralytics/yolo/utils/__init__.py", line 125, in __str__
CUDA error: misaligned address
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
... more crash stuff from other threads ...
```

**Comment by AyushExel** [comment]
This seems like a CUDA problem or something else. I don't think this is related to shared state but if it is then we might have a bug. 
I'm not sure but maybe this approach might help https://stackoverflow.com/questions/61867541/parallel-gpu-execution-with-multiple-darknet-models-in-python

**Comment by ArshadIram** [comment]
Hi, I'm having the same issue. I'm trying to use multiple webcams using different threads it works for about 1 minute without lag and with detections on GPU but then crashes with the exact same error. I then have to restart the entire development environment to run it again as it won't let me run without a restart of the application.
"RuntimeError: CUDA error: misaligned address
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions."

My assumptions were that it could be a memory issue but when running and monitoring the GPU and CPU performance and ram everything seems fine. I've read a bit online and the same error has crept up when using multiple cams in terms of classes it expects a certain value but receives double the amount of classes whcih throws the error. That's my assumption but any help on this issue would be greatly appreciated.

**Comment by glenn-jocher** [comment]
@ArshadIram i understand that you are having trouble running multiple webcams using different threads. It's not uncommon to encounter issues when working with multithreading, especially when it involves GPU resources that require careful management.

Based on the error message you're seeing ("RuntimeError: CUDA error: misaligned address"), it indeed suggests there's a problem accessing GPU memory. However, this error could be a side effect of an earlier issue rather than the cause of your problem, as the CUDA error is asynchronously reported.

The deviation of the expected number of classes causing the error makes sense. YOLOv8 processes each frame as a batch, predicting the number of objects (number of classes) present in the frame. If it receives double the expected input, this could cause an unexpected memory allocation error.

I would suggest validating input to the prediction method and ensure the right number of classes are being passed. If the error persists, you might want to consider running the model in a separate process instead of a thread, and using inter-process communication to pass data to/from your main application. Python's multiprocessing package, for example, could be used for this purpose.

Also, you might want to try to handle the cleanup of your threads, including making sure all data on the GPU is properly released once a thread is terminated. This could potentially alleviate the restart necessity upon crashing.

In summary, here are the recommended steps:

1. Validate the input to your prediction method (ensure it's not passing double the number of expected classes).
2. Consider using multiprocessing instead of multithreading, as it could handle GPU resources better.
3. Ensure proper cleanup after thread termination. 

Hope this could help you troubleshoot the issue.

**Comment by ArshadIram** [comment]
@glenn-jocher I really appreciate your detailed response to the problem we were facing. It helps a lot in understanding and solving the issue.

**Comment by glenn-jocher** [comment]
@ArshadIram i'm glad to hear that you found the previous response helpful for your problem. Your appreciation means a lot to us. Remember, our aim is to assist and ensure that you can make the most out of YOLOv8. Please don't hesitate to reach out if you have more questions or face any other issues. We're always here and happy to help. Happy coding!

**Comment by WangZhouX** [comment]
@glenn-jocher   I am using multiple threads to handle target tracking, will the model be reloaded if the thread restarts? I don't know why, but my gpu keeps growing , what should I do ?
